src/rust/engine/nailgun/src/server.rs

```diff
@@ -0,0 +1,355 @@
+      .await
+      .map_err(|e| format!("Could not bind to port {}: {:?}", port_requested, e))?;
+    let port_actual = listener
+      .local_addr()
+      .map_err(|e| format!("No local address for listener: {:?}", e))?
+      .port();
+
+    // NB: The C client requires noisy_stdin (see the `nails` crate for more info), but neither
+    // `nails` nor the pants python client do.
+    let config = nails::Config::default().noisy_stdin(false);
+    let nail = RawFdNail {
+      executor: executor.clone(),
+      runner: Arc::new(runner),
+    };
+
+    let (exited_sender, exited_receiver) = oneshot::channel();
+    let (exit_sender, exit_receiver) = oneshot::channel();
+
+    let _join = executor.spawn(Self::serve(
+      executor.clone(),
+      config,
+      nail,
+      exit_receiver,
+      exited_sender,
+      listener,
+    ));
+
+    Ok(Server {
+      exit_sender,
+      exited_receiver,
+      port: port_actual,
+    })
+  }
+
+  ///
+  /// The main loop of the server. Public for testing.
+  ///
+  pub(crate) async fn serve(
+    executor: Executor,
+    config: nails::Config,
+    nail: impl Nail,
+    should_exit: oneshot::Receiver<()>,
+    exited: oneshot::Sender<Result<(), String>>,
+    listener: TcpListener,
+  ) {
+    let exit_result = Self::accept_loop(executor, config, nail, should_exit, listener).await;
+    info!("Server exiting with {:?}", exit_result);
+    let _ = exited.send(exit_result);
+  }
+
+  async fn accept_loop(
+    executor: Executor,
+    config: nails::Config,
+    nail: impl Nail,
+    mut should_exit: oneshot::Receiver<()>,
+    mut listener: TcpListener,
+  ) -> Result<(), String> {
+    // While connections are ongoing, they acquire `read`; before shutting down, the server
+    // acquires `write`.
+    let ongoing_connections = Arc::new(RwLock::new(()));
+
+    let result = loop {
+      let tcp_stream = match future::select(listener.accept().boxed(), should_exit).await {
+        future::Either::Left((Ok((tcp_stream, _addr)), s_e)) => {
+          // Got a connection.
+          should_exit = s_e;
+          tcp_stream
+        }
+        future::Either::Left((Err(e), _)) => {
+          break Err(format!("Server failed to accept connections: {}", e));
+        }
+        future::Either::Right((_, _)) => {
+          break Ok(());
+        }
+      };
+
+      debug!("Accepted connection: {:?}", tcp_stream);
+
+      // There is a slightly delicate dance here: we wait for a connection to have acquired the
+      // ongoing connections lock before proceeding to the next iteration of the loop. This
+      // prevents us from observing an empty lock and exiting before the connection has actually
+      // acquired it. Unfortunately we cannot acquire the lock in this thread and then send the
+      // guard to the other thread due to its lifetime bounds.
+      let connection_started = Arc::new(Notify::new());
+      let _join = executor.spawn({
```

Comment:
Same as above, feel free to point me to docs if it's obvious, but where does future (the one in `_join`) get awaited? Or do we not have to worry about that somehow?